credential-verification-service/src/types.rs

```
@@ -26,37 +27,54 @@ pub struct Service {
     pub transaction_expiry_secs: u32,
 }
 
+/// Extractor with build in error handling. Like [axum::Json](Json) but will use [`RejectionError`] for rejection errors
+#[derive(FromRequest)]
+#[from_request(via(axum::Json), rejection(RejectionError))]
+#[allow(dead_code)]
+pub struct AppJson<T>(pub T);
+
+/// Error returned by REST endpoint handlers. Will
+/// be mapped to the right HTTP response (HTTP code and custom
+/// error body) by the axum middleware
 #[derive(Debug, thiserror::Error)]
 pub enum ServerError {
-    #[error("Unable to submit anchor transaction on chain successfully: {0}.")]
-    SubmitAnchorTransaction(#[from] CreateAnchorError),
-    #[error("Unable to submit anchor transaction on chain: {0}.")]
-    PresentationVerifificationFailed(#[from] VerifyError),
-    #[error("Unable to submit anchor transaction on chain: {0}.")]
-    QueryError(#[from] QueryError),
+    #[error("{0:#}")]
+    Anyhow(#[from] anyhow::Error),
```

This `ServerError` is mapped to the `StatusCode::INTERNAL_SERVER_ERROR`. Such errors could reveal internal/infra set-up (e.g. file paths to wallets, database connection string if a database would be used, ...). As a rule of thumb, I would keep the `internal error message` returned to the `client` minimalistic to only `internal error` as the client can not fix the error anyway. The logging in the service on the other hand should be verbose to simplify debugging `tracing::error!("internal error: {err}")`.

---

The problem is that right now, we really don't distinguish properly between internal server errors and client errors (like an anchor that does not exist on chain, credentials does not exists on chain etc). If we implement that distinguishing, then I agree. But for now, when we don't distinguish, I think it is better to return the error always.